docs/signature-layout.md

```diff
@@ -0,0 +1,35 @@
+# Signature File Layout
+
+A common file layout for storing and serving signatures provides a consistent way to reference image signatures. Signatures on a filesystem or a web server shall use this common layout. Signatures stored in a REST API are not required to use this common layout.
+
+## Specification
+
+This specification relies on [RFC3986](https://tools.ietf.org/html/rfc3986), focusing on defining a [path component](https://tools.ietf.org/html/rfc3986#section-3.3) to compose a concise URI reference to a signature.
+
+**SCHEME[AUTHORITY]/PATH_PREFIX/IMAGE@MANIFEST_DIGEST/signature-INT**
+
+**Definitions**
+
+* **SCHEME**: URI scheme per [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.1), e.g. **file://** or **https://**
+* **AUTHORITY**: An optional authority reference per [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.2), e.g. **example.com**
+* **PATH_PREFIX**: An arbitrary base path to the image component
+* **IMAGE**: The name of the image per [v2 API](https://docs.docker.com/registry/spec/api/#/overview). This would typically take the form of registry/repository/image but is not required to have exactly three parts. There is no requirement to include a **:PORT** component but it should be included if part of the image reference.
```

Comment:
On Thu, Nov 03, 2016 at 12:05:36PM -0700, Michael Hrivnak wrote:
> As a user doing the above, I want to mirror that content to disconnected facilities, where content must be walked in on a disk.
>
> I think those cover the basic requirements I'm hoping we can support, although obviously there are variations. As you can see, a caching proxy would not be sufficient.
>
> At each step in these use cases, a local client needs to figure out how to access an image and any associated signatures.

I think you *can* do this with a caching proxy, using @mtrmac's “[manifest+layers+signatures as single atomic unit][1]” paradigm.  You have your central, in-house registry as a caching proxy for the wider world (and maybe configure it to not push new content back to the world).  You do your management on that central registry.  Individual components within your infrastructure have their own production registries, and you bring content to them with a workflow like:

1. Image foo-bar:1.0 is vetted by QA, and gets the QA sig in the central store.
2. Setup a proxy pulling from your central store which caches into a local scratch directory.
3. Perform the operation you'd like to support (e.g. fetching the image, verifying the signatures, and walking any referenced CAS DAGs) on your proxy from (2).
4. Pack the scratch directory created in (2) and filled by (3) up and transmit it to the appropriate production registry (e.g. via a sneakernet).
5. Push the new content from the transmitted cache into the production registry.

You can get more performant if you track which objects are in the production registries so you can skip those when pushing (roughly like Git with it's remote references).

[1]: https://github.com/containers/image/pull/120#discussion_r85216586